### PR TITLE
[WIP] add examples

### PR DESCRIPTION
**Pull request by radekosmulski**
This is a WIP branch for adding examples. I have added 01 (and a brief README) and am looking for feedback if I am moving in the right direction here.

Thank you for all your help! 🙂

### CODE REVIEW

File: examples/01a-Getting-started.ipynb [hunk 1]
```diff
@@ -0,0 +1,632 @@
+{
```
Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>Can we not use special formatting for Merlin dataloader in the headline? We haven't done that in any other example. We do not use special formatting in the rest of the notebook.</p><p> </p><p>We do not use NVTabular in the notebook, so we should not confuse the reader in the beginning with it. We should remove it.</p><p> </p><p>The Overview should be something like:</p><p> </p><p>"Merlin dataloader is a library for constructing highly optimized dataloaders to accelerate training pipelines in TensorFlow (Keras) and PyTorch. In this example, we will provide a simple pipeline to train a MatrixFactorization Model in TensorFlow with Merlin dataloader based on the MovieLens dataset.</p><p> </p><p>The core features of Merlin dataloader:</p><p>- Accelerate pipelines by upto 10x compared to other dataloaders</p><p>- Handles larger than memory dataset by streaming data from disk</p><p>- Support for common data formats: CSV, Parquet, Avro</p><p>- Distributed training support</p><p> </p><p>Learning Objectives:</p><p>- Using Merlin dataloader to train a TensorFlow Keras Model</p><p>"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>I think we should keep the description of MovieLens short. The focus is on the dataloader and Movielens is our "hello world" example.</p><p> </p><p>I would call the headline "Downloading and perparing the dataset".</p><p> </p><p>I would just say that we use Movielens as an example how to use Merlin dataloaders.</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>In general, can we make the example more lean with focus on the dataloader + training? </p><p> </p><p> </p><p>It seems that we can use merlin.core -&gt; we have a function to download and extract a dataset:</p><pre class="ql-syntax" spellcheck="false">from merlin.core.utils import download_file
</pre><p> </p><p>Can we use this? See</p><p>https://github.com/NVIDIA-Merlin/NVTabular/blob/main/examples/getting-started-movielens/01-Download-Convert.ipynb</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>Do we need to install unzip? How does it work in this example: </p><p>https://github.com/NVIDIA-Merlin/NVTabular/blob/main/examples/getting-started-movielens/01-Download-Convert.ipynb </p><p> </p><p>I think we should not install libraries in the example. We should make a note, that it requires unzip (maybe download_file function doesnt rely on unzip)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>We can use a utils function and do "Categorify" other preprocessing under the hood and store it to parquet. (If we want)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>We miss another headline - so far we have only "Downloading the datase" and everything is in that headline.</p><p> </p><p>I think we need to add a new one here: Training a TensorFlow Keras Model with Merlin dataloader</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>Can we have all required imports in the beginning of the notebook?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>It seems that the dataloader does not support cat_names, cont_names and label_names - is that correct?</p><p>It seems it relies on a data schema -&gt; Either, we need to provide an easy tool to define a schema manually or add the 3 parameters back?</p><p> </p><p>Can we make this more generic/explanable. E.g.</p><p> </p><p>label_columns = ['rating']</p><p>def process_batch(data, _):</p><p>x = {col: data[col] for col in data.keys() if col not in label_columns}</p><p>y = data[label_columns]</p><p>&nbsp;&nbsp;return (x, y)</p><p> </p><p>"What <code>Tensorflow</code> expects to see are targets as the 2nd position in the tuple. " -&gt; Let's rewrite it to make it more clear - something like</p><p>"TensorFlow Kera's .fit function expects the data to get as a tuple (x, y), with x are the input features and y is the label. We need to provide this information to the dataloader. We can add a custom function to convert the data into the tuple with <code>process_batch</code>"</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>Lets remove this. This is pretty confusing :) We would not validate without training?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>Do you know why we get the warnings/errors?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (bschifferer) [inline examples/01a-Getting-started.ipynb:1]: <p>We should summarize the end of the notebook with something like:</p><p> </p><pre class="ql-syntax" spellcheck="false">## Conclusion
We demonstrated how to train a TensorFlow Keras model with Merlin dataloader. Merlin dataloader can accelerate existing TensorFlow pipelines with minimal code changes. 

## Next Steps
Merlin dataloader is part of NVIDIA Merlin, a open source framework for recommender systems. In this example, we looked only on a specific use-case to accelerate existing training pipelines. We provide more libraries to make recommender system pipelines easier:
- NVTabular is a library to accelerate and scale feature engineering
- Merlin Models is a library with high-quality implementations of popular recommender systems architectures 

The libraries are designed to work closely togethes. We recommend to checkout our exmaples:
- Getting Started with NVTabular: Process Tabular Data On GPU
- Getting Started with Merlin Models: Develop a Model for MovieLens 

In the example, From ETL to Training RecSys models - NVTabular and Merlin Models integrated example, we explain how the close collaboration works.
</pre><p> </p><p>Can you add links for the libraries and examples?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

File: examples/01b-Getting-started-Tensorflow.ipynb [hunk 2]
```diff
@@ -0,0 +1,577 @@
+{
```
Comment (bschifferer) [inline examples/01b-Getting-started-Tensorflow.ipynb:1]: <p>in my understanding, this example is similar to examples/01a-Getting-started.ipynb except of using NVTabular. I think we should reference NVTabular in the end and provide two examples (TensorFlow/PyTorch) with focus on dataloader.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: examples/01c-Getting-started-Pytorch.ipynb [hunk 3]
```diff
@@ -0,0 +1,553 @@
+{
```
Comment (bschifferer) [inline examples/01c-Getting-started-Pytorch.ipynb:1]: <p>Same Feedback as 01-Getting-Started </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (bschifferer) [inline examples/01c-Getting-started-Pytorch.ipynb:1]: <p>We should use a minimal example without NVTabular. We can build one utils function to download and process the dataset?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (bschifferer) [inline examples/01c-Getting-started-Pytorch.ipynb:1]: <p>lets call the function evaluate -&gt; calculate_loss could be single batch or full batch?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (bschifferer) [inline examples/01c-Getting-started-Pytorch.ipynb:1]: <p>Let's do not calculate loss in the beginning, I havent seen that in another pipeline/example?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (bschifferer) [inline examples/01c-Getting-started-Pytorch.ipynb:1]: <p>Similar conclusion and next steps as getting started - however, we cannot reference to merlin models. We can reference other examples such as</p><p> </p><p>https://github.com/NVIDIA-Merlin/Merlin/blob/main/examples/getting-started-movielens/03-Training-with-PyTorch.ipynb</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

**Comment by bschifferer** [comment]
It looks good - I added some comments.

We need a unittest for the final notebooks

File: examples/01a-Getting-started.ipynb (hunk 1 above)
Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>changes made</p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>done</p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>absolutely, that is a great point, switched to using <code>download_file</code></p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>everything works now with just <code>download_file</code> without a need to install anything else 👍 </p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>works without any additional preprocessing so in the interest of keeping it simple and not dragging reader's attention from the dataloader would just keep the data as is</p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>good point, will pay more attentiont to those headlines going forward</p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>this reads much better now and communicates much more to the reader 🙂👍</p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>sure thing, removed</p><p> </p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>I am unforutnately not sure why these do come up. I googled for it but the answer I was able to find is that this is something that started to occur with moving to 2.9, but no explanation as to how to get rid of them</p><p> </p>

Comment (radekosmulski) [reply examples/01a-Getting-started.ipynb:1]: <p>absolutely, done! 🙂🙌</p>

File: examples/01b-Getting-started-Tensorflow.ipynb (hunk 2 above)
Comment (radekosmulski) [reply examples/01b-Getting-started-Tensorflow.ipynb:1]: <p>agreed!</p>

File: examples/01c-Getting-started-Pytorch.ipynb (hunk 3 above)
Comment (radekosmulski) [reply examples/01c-Getting-started-Pytorch.ipynb:1]: <p>done</p>

Comment (radekosmulski) [reply examples/01c-Getting-started-Pytorch.ipynb:1]: <p>removed the dependency on nvtabular completely, having this mirror the previous tensorflow notebook</p>

Comment (radekosmulski) [reply examples/01c-Getting-started-Pytorch.ipynb:1]: <p>good point, made the change!</p>

Comment (radekosmulski) [reply examples/01c-Getting-started-Pytorch.ipynb:1]: <p>removed</p>

File: examples/01b-Getting-started-Pytorch.ipynb [hunk 4]
```diff
@@ -0,0 +1,487 @@
+{
```
Comment (bschifferer) [inline examples/01b-Getting-started-Pytorch.ipynb:1]: <p>maybe we should add in the headline in each example the framework</p><p> </p><p>Getting Started with Merlin dataloader and PyTorch </p><p>Getting Started with Merlin dataloader and TensorFlow</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (bschifferer) [inline examples/01b-Getting-started-Pytorch.ipynb:1]: <p>Maybe add a sentence why you print this out (or we remove it :) )</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (bschifferer) [inline examples/01b-Getting-started-Pytorch.ipynb:1]: <p>We do not calculate it anymore, do we?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/dataloader/pull/47/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (radekosmulski) [reply examples/01b-Getting-started-Pytorch.ipynb:1]: <p>made the change 👍</p>

Comment (radekosmulski) [reply examples/01b-Getting-started-Pytorch.ipynb:1]: <p>yes, thank you for spotting it, my bad for not removing it, removed it now</p>

**Comment by radekosmulski** [comment]
rerun tests
